source/extensions/filters/network/connection_limit/connection_limit.cc

```diff
@@ -0,0 +1,113 @@
+#include "source/extensions/filters/network/connection_limit/connection_limit.h"
+
+#include "envoy/extensions/filters/network/connection_limit/v3/connection_limit.pb.h"
+
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ConnectionLimitFilter {
+
+Config::Config(
+    const envoy::extensions::filters::network::connection_limit::v3::ConnectionLimit& proto_config,
+    Stats::Scope& scope, Runtime::Loader& runtime)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)),
+      max_connections_(PROTOBUF_GET_WRAPPED_REQUIRED(proto_config, max_connections)),
+      connections_(0), delay_(PROTOBUF_GET_OPTIONAL_MS(proto_config, delay)) {}
+
+ConnectionLimitStats Config::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  const std::string final_prefix = "connection_limit." + prefix;
+  return {ALL_CONNECTION_LIMIT_STATS(POOL_COUNTER_PREFIX(scope, final_prefix),
+                                     POOL_GAUGE_PREFIX(scope, final_prefix))};
+}
+
+bool Config::incrementConnectionWithinLimit() {
+  auto conns = connections_.load(std::memory_order_relaxed);
+  while (conns < max_connections_) {
+    // Testing hook.
+    synchronizer_.syncPoint("increment_pre_cas");
+
+    if (connections_.compare_exchange_weak(conns, conns + 1, std::memory_order_release,
+                                           std::memory_order_relaxed)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void Config::incrementConnection() { connections_++; }
+
+void Config::decrementConnection() {
+  auto conns = connections_.load(std::memory_order_relaxed);
+  do {
+    ASSERT(conns > 0);
+
+    // Loop while the weak CAS fails trying to update the conns value.
+  } while (!connections_.compare_exchange_weak(conns, conns - 1, std::memory_order_release,
```

Comment:
I don't think you need a CAS here? Just do an atomic `connections--`?

Reply:
Make sense, updated.